gst/tensor_ros_src/tensor_ros_src.cc

```diff
@@ -388,6 +398,80 @@ gst_tensor_ros_src_dispose (GObject * object)
   G_OBJECT_CLASS (parent_class)->dispose (object);
 }
 
+/**
+ * @brief Start handler
+ */
+static gboolean
+gst_tensor_ros_src_start (GstBaseSrc * src)
+{
+  GstTensorRosSrc *rossrc = GST_TENSOR_ROS_SRC (src);
+
+  silent_debug (rossrc, "start RosSubscriber");
+  rossrc->ros_sub = new TensorRosSub ("TensorRosSub", rossrc->topic_name,
+    rossrc, rossrc->freq_rate);
+  rossrc->ros_sub->Start (&rossrc->thread);
+
+  while (!rossrc->configured)
+    g_usleep (G_USEC_PER_SEC / rossrc->freq_rate);
+
+  silent_debug (rossrc, "tensor configuration is completed");
+  return TRUE;
+}
+
+/**
+ * @brief Stop handler
+ */
+static gboolean
+gst_tensor_ros_src_stop (GstBaseSrc * src)
+{
+  GstTensorRosSrc *rossrc = GST_TENSOR_ROS_SRC (src);
+
+  silent_debug (rossrc, "stop RosSubscriber");
+  rossrc->ros_sub->RequestStop ();
+  return TRUE;
+}
+
+/**
+ * @brief Get caps handler
+ */
+static GstCaps *
+gst_tensor_ros_src_get_caps (GstBaseSrc * src, GstCaps * filter)
+{
+  GstTensorRosSrc *rossrc = GST_TENSOR_ROS_SRC (src);
+  GstCaps *caps;
+  GstCaps *ret_caps;
+
+  caps = gst_tensor_caps_from_config (&rossrc->config);
```

Comment:
IMO, getting the caps from config and updating should be done in _fixate() / _negotiate()
using the caps from config returns caps which have not been set on the pad, which might lead to unexpected behavior

I recommend to use `gst_pad_get_current_caps (src->srcpad)` here and get caps from tensor in _fixate().

Reply:
@kparichay Is gst_tensor_caps_from_config( ) possible to return caps that is not set yet? IMO, in that case it returns NULL and we just need to check whether returned caps NULL or NOT.

Reply:
@again4you @kparichay For the record, we discussed this issue. Thanks!

Reply:
@kparichay  @wooksong Your comments are really helpful for me. Thanks!

Reply:
> @kparichay Is gst_tensor_caps_from_config( ) possible to return caps that is not set yet? IMO, in that case it returns NULL and we just need to check whether returned caps NULL or NOT.

`gst_tensor_caps_from_config()` returns the caps from the current config. Those caps might be or might not be set on the caps, depends on when it is being called.

Reply:
@again4you  @wooksong The discussion was helpful to me as well, Thanks.